dragonfly_energy/properties/model.py

```diff
@@ -0,0 +1,391 @@
+        if len(duplicate_names) != 0:
+            if raise_exception:
+                raise ValueError(
+                    'The model has the following duplicated ConstructionSet '
+                    'names:\n{}'.format('\n'.join(duplicate_names)))
+            return False
+        return True
+
+    def check_duplicate_program_type_names(self, raise_exception=True):
+        """Check that there are no duplicate ProgramType names in the model."""
+        p_type_names = set()
+        duplicate_names = set()
+        for p_type in self.program_types:
+            if p_type.name not in p_type_names:
+                p_type_names.add(p_type.name)
+            else:
+                duplicate_names.add(p_type.name)
+        if len(duplicate_names) != 0:
+            if raise_exception:
+                raise ValueError(
+                    'The model has the following duplicated ProgramType '
+                    'names:\n{}'.format('\n'.join(duplicate_names)))
+            return False
+        return True
+    
+    def apply_properties_from_dict(self, data):
+        """Apply the energy properties of a dictionary to the host Model of this object.
+
+        Args:
+            data: A dictionary representation of an entire honeybee-core Model.
+                Note that this dictionary must have ModelEnergyProperties in order
+                for this method to successfully apply the energy properties.
+        """
+        assert 'energy' in data['properties'], \
+            'Dictionary possesses no ModelEnergyProperties.'
+
+        # set the terrain
+        if 'terrain_type' in data['properties']['energy']:
+            self.terrain_type = data['properties']['energy']['terrain_type']
+
+        materials, constructions, construction_sets, schedule_type_limits, \
+            schedules, program_types = \
+            hb_model_properties.ModelEnergyProperties.load_properties_from_dict(data)
+
+        # collect lists of energy property dictionaries
+        building_e_dicts, story_e_dicts, room2d_e_dicts, context_e_dicts = \
+            model_extension_dicts(data, 'energy')
+
+        # apply energy properties to objects uwing the energy property dictionaries
+        for bldg, b_dict in zip(self.host.buildings, building_e_dicts):
+            bldg.properties.energy.apply_properties_from_dict(b_dict, construction_sets)
+        for story, s_dict in zip(self.host.stories, story_e_dicts):
+            story.properties.energy.apply_properties_from_dict(s_dict, construction_sets)
+        for room, r_dict in zip(self.host.rooms, room2d_e_dicts):
+            room.properties.energy.apply_properties_from_dict(
+                r_dict, construction_sets, program_types)
+        for shade, s_dict in zip(self.host.shades, context_e_dicts):
+            shade.properties.energy.apply_properties_from_dict(
+                s_dict, constructions, schedules)
+
+    def to_dict(self, include_global_construction_set=True):
```

Comment:
Some of the to_dict methods are getting pretty long. I wonder if we can break it down into smaller pieces and then call them here. It might also be useful for debugging later if you only want to debug part of the to_dict for one or a number of the keys.

Reply:
That's a fair suggestion. Some parts of the method can't be separated out because they depend on previous steps but I think I can pull out at least 2 large methods that could get called from `to_dict`.

Reply:
Done!